examples/component.rs

```diff
@@ -0,0 +1,166 @@
+extern crate env_logger;
+extern crate log;
+extern crate z3;
+
+use z3::{
+    ast::{Ast, Bool, Datatype, Int, BV},
+    Config, Context, DatatypeBuilder, DatatypeSort, SatResult, Solver,
+};
+
+fn main() {
+    let _ = env_logger::try_init();
+    let cfg = Config::new();
+    let ctx = Context::new(&cfg);
+
+    // Define a new datatype named Instruction which has 3 variants.
+    // 1. LocalGet 0
+    // 2. I32Add
+    // 3. Nop
+    let components = DatatypeBuilder::new(&ctx)
+        .variant("LocalGet0", &[])
+        .variant("I32Add", &[])
+        .variant("Nop", &[])
+        .finish("Instruction");
+
+    // Function that converts an Instruction datatype into z3::ast::Int which represents the number
+    // of values the instruction pops off from the stack.
+    fn pop_cnts<'a>(
+        ctx: &'a Context,
+        components: &DatatypeSort<'a>,
+        instr: &Datatype<'a>,
+    ) -> Int<'a> {
+        components.variants[0]
+            .tester
+            .apply(&[&instr.clone().into()])
+            .as_bool()
+            .unwrap()
+            .ite(
+                &(Int::from_u64(&ctx, 0)),
+                &components.variants[1]
+                    .tester
+                    .apply(&[&instr.clone().into()])
+                    .as_bool()
+                    .unwrap()
+                    .ite(&(Int::from_u64(&ctx, 2)), &(Int::from_u64(&ctx, 0))),
+            )
+    }
+
+    // Function that converts an Instruction datatype into z3::ast::Int which represents the number
+    // of values the instruction pushes to the stack.
+    fn push_cnts<'a>(
+        ctx: &'a Context,
+        components: &DatatypeSort<'a>,
+        instr: &Datatype<'a>,
+    ) -> Int<'a> {
+        components.variants[0]
```

Comment:
Can we write push_cnts and pop_cnts in a more stylized form? For instance,
if (instruction == 0) return ...
else if (instruction == 1) return ...
else if (instruction == 2) return ...
...

Right now, it's not clear what the push_cnts for a particular component type is. Or better still create a map or dictionary to hold this information and use the component/instruction id as the key to lookup the map.

Reply:
No, it can't be written like that.  See the function instr_simplified below. The same argument holds for push_cnts() and pop_cnts() too. 

First of all, the variable `instr`, z3 representation of instruction, is of type [z3::ast::Datatype](https://docs.rs/z3/0.4.0/z3/ast/struct.Datatype.html). If you look at its methods, it doesn't provide a function to get its id or string representation. Even if it did, for the same reason as instr_simplified, it won't work, because instr is never instantiated to a concrete value, it's just a symbolic representation of an instruction. 

Still, it's possible to use a map from datatype to unsigned ints as in pop_cnts_map() function. However, you can't get away from the ugly 
`components.variants[0].constructor.apply(&[]).as_datatype().unwrap()` to create instructions.